appengine/pond_storage/opponents.py

```diff
@@ -19,29 +19,179 @@
 
 import cgi
 import json
+import math
+import random
 from google.appengine.ext import ndb
 from pond_storage import *
 
-def get_opponent_ducks(duck):
-  """Returns list of opponent ducks."""
-  # TODO: Add filter for published ducks
-  # TODO: Add logic here to pick the duck based on the ranking
-  user_key = get_user_key(users.get_current_user())
-  duck_query = Duck.query()
-  duck_list = []
-  for duck in duck_query:
-    if duck.key.parent() != user_key:
-      duck_list.append(get_duck_info(duck))
-  return duck_list[0:3]
+_START_PERCENT = .10
+"""Looks above or below the current ranking by this number to find the start ranking."""
+
+_END_PERCENT = .15
+"""Looks above or below the current ranking by this number to find the end ranking."""
+
+_NEARBY_PERCENT = .10
+"""Look above and below this percent to find nearby entries."""
+
+_OPPONENTS_NEEDED = 3
+"""The number of ducks opponents needed"""
+
+def get_entries_in_range(user_entry, exclude_entries, start_rank, end_rank):
+  """Get all entries in the given range.
+  Args:
+    user_entry: The current user's LeaderboardEntry.
+    exclude_entries: A list of LeaderboardEntry values already used.
+    start_rank: The number of the rank we want to start getting LeaderboardEntrys at.
+    end_rank: The number of the rank we want to stop getting LeaderboardEntrys at.
+  Returns:
+    A list of LeaderboardEntrys.
+  """
+  new_entries = LeaderboardEntry.query(
+    LeaderboardEntry.leaderboard_key == user_entry.leaderboard_key,
+    LeaderboardEntry.ranking >= start_rank,
+    LeaderboardEntry.ranking <= end_rank
+  ).fetch(10)
+  new_entries = [x for x in new_entries if x not in exclude_entries and x.key != user_entry.key]
+
+  return new_entries
+
+def get_nearby_entries(user_entry, entries, total_entries, num_needed):
+  """Get ducks with similar rankings to the users duck
+  Args:
+    user_entry: The current user's LeaderboardEntry
+    entries: A list of LeaderboardEntry values already used
```

Comment:
please update entries > exclude_entries here as well as get_worse_entry and get better_entry to match get_entries_in_range